packages/xpath/src/functions/xforms/node-set.ts

```diff
@@ -32,6 +34,221 @@ export const countNonEmpty = new NumberFunction(
+		// understanding was clarified in review of that orignal effort, and is
+		// borne out by new tests exercising depth > 1, which demonstrate the same
+		// behavior in JavaRosa.
+		//
+		// Note: we start iterating here at index 1 so assertions related to
+		// positional argument index are clear.
+		for (let i = 1; i < args.length; i += 2) {
+			const repeatsArg = args[i];
+			const positionArg = args[i + 1];
+
+			assertArgument(i, repeatsArg);
+			assertArgument(i + 1, positionArg);
+
+			// Evaluate position first, because...
+			const position = positionArg.evaluate(context).toNumber();
+
+			// ... if any "index" (position) is `NaN`, we short-circuit. This is
+			// expected behavior because the equivalent `/data/repN[posN]/target`
+			// expression would do the same.
+			if (Number.isNaN(position)) {
+				return [];
+			}
+
+			// Reiterating the point made describing this loop for future clarity:
+			// this collects **all** of the nodes matching the `repeatN` expression.
+			// We filter them in a later step.
+			const repeats = evaluateArgumentToFilterableNodes(context, repeatsArg);
+
+			// No repeats = nothing to "index" = short circuit
+			if (repeats.length === 0) {
+				return [];
+			}
+
+			pairs.push({
+				repeats,
+				position,
+			});
+		}
+
+		// Sort the results of each `repeatN`/`indexN` pair, by containment order.
+		//
+		// Note: the `repeatN`/`indexN` pairs can be supplied in any order (this is
+		// consistent with behavior in JavaRosa, likely as a side effect of the
+		// function being implemented there by transforming the expression to its
+		// LocationPath equivalent).
+		pairs = pairs.sort(compareContainmentDepth);
+
+		// Resolve repeats at the specified/evaluated position, in document depth
+		// order by:
+		//
+		// 1. Filtering each set of repeats to include **only** the nodes contained
+		//    by the previously resolved repeat.
+		//
+		// 2. Selecting the repeat at the specified/evaluated position (of those
+		//    filtered in 1).
+		let repeatContextNode: ContextNode;
+
+		for (const pair of pairs) {
+			const { position } = pair;
+			const repeats = pair.repeats.filter((repeat) => {
+				return repeatContextNode?.contains(repeat) !== false;
+			});
+
+			// Select next repeat context at the current `repeatN`/`indexN` position.
+			//
+			// Note: despite terminology used in the spec, `indexN` is treated as
+			// equivalent to an XPath position predicate: it is 1-based.
+			const positionedRepeat = repeats[position - 1];
+
+			// No repeat context is found = nothing to target = short-circuit
+			if (positionedRepeat == null) {
+				return [];
+			}
+
+			repeatContextNode = positionedRepeat;
+		}
+
+		// Resolve **all** target nodes.
+		const targetNodes = evaluateArgumentToFilterableNodes(context, target);
+
+		// Filter only the target nodes contained by the deepest repeat context node.
+		return targetNodes.filter((targetNode) => {
```

Comment:
I'll refer to the casting details in [my comment the other thread](https://github.com/getodk/web-forms/pull/195#discussion_r1733604219), with emphasis on how node-sets are cast to strings. This is also how XPath semantics addresses node-set values.

Anyway, as I said in that thread: I'm open to making this less capable if that's what we decide we really want.

But to be clear, a `NodeSetFunction` implementation can't return a single node, it must return `Iterable<Node>`. So if we decide that the less capable thing we're after is guaranteeing single node-ness (rather than the value-ness, which is already handled by casting up the stack), we'd still need to either rewrap the single node result in an array (empty if null) or convert this to a generator and `yield` the non-null result.